File: sdk/@launchdarkly/observability-node/src/otel/sampling/sampleSpans.ts
import { ReadableSpan } from '@opentelemetry/sdk-trace-base'
import { Attributes } from '@opentelemetry/api'
import { merge } from '@opentelemetry/core'
import { ExportSampler } from './ExportSampler'

export function cloneReadableSpanWithAttributes(
	span: ReadableSpan,
	attributes: Attributes,
): ReadableSpan {
	const spanContext = span.spanContext()
	const cloned = {
		name: span.name,
		kind: span.kind,
		spanContext: () => spanContext,
		parentSpanContext: span.parentSpanContext,
		startTime: span.startTime,
		endTime: span.endTime,
		status: span.status,
		attributes: merge(span.attributes, attributes),
		links: span.links,
		events: span.events,
		duration: span.duration,
		ended: span.ended,
		resource: span.resource,
		instrumentationScope: span.instrumentationScope,
		droppedAttributesCount: span.droppedAttributesCount,
		droppedEventsCount: span.droppedEventsCount,
		droppedLinksCount: span.droppedLinksCount,
	}
	return cloned
}

export function sampleSpans(
	items: ReadableSpan[],
	sampler: ExportSampler,
): ReadableSpan[] {
	if (!sampler.isSamplingEnabled()) {
		return items
	}
	const omittedSpansIds: string[] = []
	const spanById: Record<string, ReadableSpan> = {}
	const childrenByParentId: Record<string, string[]> = {}

	// The first pass we sample items which are directly impacted by a sampling decision.
	// We also build a map of children spans by parent span id, which allows us to quickly traverse the span tree.
	for (const item of items) {
		if (item.parentSpanContext?.spanId) {
			childrenByParentId[item.parentSpanContext.spanId] =
				childrenByParentId[item.parentSpanContext.spanId] || []
			childrenByParentId[item.parentSpanContext.spanId].push(
				item.spanContext().spanId,
			)
		}
		const sampleResult = sampler.sampleSpan(item)
		if (sampleResult.sample) {
			if (sampleResult.attributes) {
				spanById[item.spanContext().spanId] =
					cloneReadableSpanWithAttributes(
						item,
						sampleResult.attributes,
					)
			} else {
				spanById[item.spanContext().spanId] = item
			}
		} else {
			omittedSpansIds.push(item.spanContext().spanId)
		}
	}

	// Find all children of spans that have been sampled out and remove them.
	// Repeat until there are no more children to remove.
	while (omittedSpansIds.length > 0) {
		const spanId = omittedSpansIds.shift()
		const affectedSpans: string[] | undefined = childrenByParentId[spanId!]
		if (!affectedSpans) {
			continue
		}

		for (const spanIdToRemove of affectedSpans) {
			delete spanById[spanIdToRemove]
			omittedSpansIds.push(spanIdToRemove)
		}
	}
	return Object.values(spanById)
}

File: sdk/@launchdarkly/observability-android/lib/src/main/kotlin/com/launchdarkly/observability/sampling/SampleSpans.kt
package com.launchdarkly.observability.sampling

import io.opentelemetry.api.common.Attributes
import io.opentelemetry.sdk.common.InstrumentationScopeInfo
import io.opentelemetry.sdk.trace.data.SpanData

fun sampleSpans(
    items: List<SpanData>,
    sampler: ExportSampler
): List<SpanData> {

    if (!sampler.isSamplingEnabled()) {
        return items
    }

    val omittedSpansIds = ArrayDeque<String>() // ArrayDeque prevents O(n²) complexity caused by removing the first element in the while loop.
    val spanById = mutableMapOf<String, SpanData>()
    val childrenByParentId = mutableMapOf<String, MutableList<String>>()

    // The first pass we sample items which are directly impacted by a sampling decision.
    // We also build a map of children spans by parent span id, which allows us to quickly traverse the span tree.
    for (item in items) {
        item.parentSpanContext?.spanId?.let { parentSpanId ->
            childrenByParentId.getOrPut(parentSpanId) { mutableListOf() }
                .add(item.spanContext.spanId)
        }
        val sampleResult = sampler.sampleSpan(item)
        if (sampleResult.sample) {
            if (sampleResult.attributes != null) {
                spanById[item.spanContext.spanId] = cloneSpanDataWithAttributes(item, sampleResult.attributes)
            } else {
                spanById[item.spanContext.spanId] = item
            }
        } else {
            omittedSpansIds.addLast(item.spanContext.spanId)
        }
    }

    // Find all children of spans that have been sampled out and remove them.
    // Repeat until there are no more children to remove.
    while (omittedSpansIds.isNotEmpty()) {
        val spanId = omittedSpansIds.removeFirst()
        val affectedSpans = childrenByParentId[spanId] ?: continue

        for (spanIdToRemove in affectedSpans) {
            spanById.remove(spanIdToRemove)
            omittedSpansIds.addLast(spanIdToRemove)
        }
    }
    return spanById.values.toList()
}

private fun cloneSpanDataWithAttributes(
    span: SpanData,
    attributes: Attributes
): SpanData {
    return object : SpanData by span {
        override fun getTraceId(): String? = span.traceId
        override fun getSpanId(): String? = span.spanId
        override fun getParentSpanId(): String? = span.parentSpanId
        override fun getInstrumentationScopeInfo(): InstrumentationScopeInfo? = span.instrumentationScopeInfo

        override fun getAttributes(): Attributes {
            return Attributes.builder().apply {
                putAll(span.attributes)
                putAll(attributes)
            }.build()
        }
    }
}
